### ISSUE TITLE
TimeSpeed: inconsistent TickLength configuration

### ISSUE THREAD

**Issue by gralewski** [Pathoschild/StardewCommunityMods#4]
Hey, another suggestion.
Currently if you don't set Indoors and Outdoors, the addon sets them in config by itself.
Probably it's not the best behaviour. Timespeed 1.0 was more consistent in expected behaviour.

Use case, let's assume what we want:
1. 10 second tick length by default.
2. Being able to change it on the fly and that newly set value stays when we change location.
3. Have 30 second tick in mines.

So we use:
  "DefaultTickLength": 10.0,
  "TickLengthByLocation": {
    "UndergroundMine": 120.0
  }

What happens?
1. Addon adds Indoors and Outdoors to config.
2. 10 second tick length gets set every time we go indoors and outdoors, so our tweaking during the game get's overrided. (Most annoying).
3. DefaultTickLength currently becomes (wrongly?) obsolete.

So in other words:
1.  I would like to start the game with 10 tick length.
2. Have the option to change it on the fly. And that it stays that way until i change location to mines or reload config to default by hotkey B.

I hope the suggetion will help to make this great thing great even more ;)

**Comment by alexb5dh** [comment]
> Addon adds Indoors and Outdoors to config.

This was added to provide some kind of 'in-place' documentation about supported location types. Since it's now stated in description I agree such behaviour should be removed.

>I would like to start the game with 10 tick length.
Have the option to change it on the fly. And that it stays that way until i change location to mines or reload config to default by hotkey B.

Saving hotkey changes for current game session or even to config would be a nice feature but I don't see how it can be implemented intuitively for player. For example, I would expect such changes to stay permanently but for this location only.

Currently changes are expected to be tuned in game, manually added to config and then reloaded. I do want to move this functionality to in-game GUI, but this requires a noticable amount of free time for me to understand how this is implemented in SV.

**Comment by gralewski** [comment]
Timespeed 1 remembers somehow tick value when changing locations.
Does the game require Timespeed 2 to read config every time the change happens?

**Comment by alexb5dh** [comment]
> Timespeed 1 remembers somehow tick value when changing locations.

Remembering change is not a problem. But this should be done in an intuitive and simple way.
The question is -"_What location(s) should be affected when changing time via in-game keys?_".

Suppose we want to remember such changes for the game session.
As an example with your config
```json
{
  "DefaultTickLength": 10.0,
  "TickLengthByLocation": {
    "UndergroundMine": 120.0
  }
}
```
and changing time in **Home** to 20.
- I think most playes would expect it to save as following:
```json
{
  "DefaultTickLength": 10.0,
  "TickLengthByLocation": {
    "UndergroundMine": 120.0,
    "Home": 20
  }
}
```
- You suggested another possible case where it should become:
 ```json
{
  "DefaultTickLength": 20.0,
  "TickLengthByLocation": {
    "UndergroundMine": 120.0,
  }
}
```

> Does the game require Timespeed 2 to read config every time the change happens?

Hotkey changes sticks only until location changes. If you want it to be permanently, you modify config manually and hit reload button to apply.
This is expected to be one-time-game-tuning operation.

**Comment by gralewski** [comment]
Well, the explanation is a bit tricky.

IMO in the mind of a person, the the subject is probably "me" not "where i am".
In another words "the feel of time speed" is the subject for the player, not the location.

So changing in Farm from 10 to 20 has to change DefaultTickLength, not just Farm.
Then the feeling is consistent for the player, the thinks he did the time in general to be 2 times slower.

So in player's mind the time is 2 time slower.
When he changes location, this feeling should be consistent, also 20 second.
Everywhere he goes has to be 20 seconds then.

What about config?
120 is override just for mines.
So he has his 20 second tick, he goes to mines and the tick is 120s here.
Changing this should change tick in this particular location now.
Player goes out, the tick gets its 20s value.
I think it's most logical explanation.

Also still /2 and *2 is more logical to me and scales better than -10/+10.
I would really be happy if i could change values of tick to values: 1s, 2s, 4s, 8s, 16s, 32s, 64s, 128s etc.
Technically it's just multiplying and dividing by 2 with 1 as minimum and for example 4096 as max.

The question is if you can achieve it technically.
Somehow timespeed 1 remembered changes to time flow during locations.
I see some addons like "NoSoilDecay" can save into mod folder, not only between locations but also between game sessions.

I just installed Visual Studio 2017, made the tutorial and compiled my first tutorial mod.
I think i'll try to learn a bit C#, then smapi and then i'll might be more of technical help.
Never modded before and never programmed C# but let's try :)

Edit: SkullCaveSaver (saves to config).
You can probably temporary store values during gameplay just as it does but reset cache to config values every time the game starts (or maybe not reset?) it might be even more fun then!

**Comment by Pathoschild** [comment]
Closing this since community mods are discontinued.

**Comment by Pathoschild** [comment]
Closing this since community mods are discontinued. All TimeSpeed changes have been merged into [the original repository](https://github.com/cantorsdust/StardewMods).
